pkg/tsdb/calculator.go

```diff
@@ -0,0 +1,226 @@
+package tsdb
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana/pkg/tsdb/interval"
+)
+
+var (
+	defaultRes         int64 = 1500
+	defaultMinInterval       = time.Millisecond * 1
+	year                     = time.Hour * 24 * 365
+	day                      = time.Hour * 24
+)
+
+type Interval struct {
+	Text  string
+	Value time.Duration
+}
+
+type intervalCalculator struct {
+	minInterval time.Duration
+}
+
+type Calculator interface {
+	Calculate(timerange backend.TimeRange, minInterval time.Duration) Interval
+}
+
+type CalculatorOptions struct {
+	MinInterval time.Duration
+}
+
+func NewCalculator(opts ...CalculatorOptions) *intervalCalculator {
+	calc := &intervalCalculator{}
+
+	for _, o := range opts {
+		if o.MinInterval == 0 {
+			calc.minInterval = defaultMinInterval
+		} else {
+			calc.minInterval = o.MinInterval
+		}
+	}
+
+	return calc
+}
+
+func (i *Interval) Milliseconds() int64 {
+	return i.Value.Nanoseconds() / int64(time.Millisecond)
+}
+
+func (ic *intervalCalculator) Calculate(timerange backend.TimeRange, minInterval time.Duration) Interval {
+	to := timerange.To.Unix()
+	from := timerange.From.Unix()
+	intrvl := time.Duration((to - from) / defaultRes)
+
+	if intrvl < minInterval {
+		return Interval{Text: interval.FormatDuration(minInterval), Value: minInterval}
+	}
+
+	rounded := roundInterval(intrvl)
+	return Interval{Text: interval.FormatDuration(rounded), Value: rounded}
+}
+
+func GetIntervalFrom(interval, timeInterval string, queryIntervalMs int64, defaultInterval time.Duration) (time.Duration, error) {
+	// intervalMs field appears in the v2 plugins API and should be preferred
+	// if 'interval' isn't present.
+	if interval == "" {
+		if queryIntervalMs != 0 {
+			return time.Duration(queryIntervalMs) * time.Millisecond, nil
+		}
+	}
+
+	if interval == "" {
+		dsInterval := timeInterval
+		if dsInterval != "" {
+			interval = dsInterval
+		}
+	}
+
+	if interval == "" {
+		return defaultInterval, nil
+	}
+
+	interval = strings.Replace(strings.Replace(interval, "<", "", 1), ">", "", 1)
+	isPureNum, err := regexp.MatchString(`^\d+$`, interval)
+	if err != nil {
+		return time.Duration(0), err
+	}
+	if isPureNum {
+		interval += "s"
+	}
+	parsedInterval, err := time.ParseDuration(interval)
+	if err != nil {
+		return time.Duration(0), err
+	}
+
+	return parsedInterval, nil
+}
```

Comment:
Changed the order of the arguments to make this API more clear to the end-user. As discussed, no longer want to move this to the SDK so maybe moving this into the `interval` package to keep all logic on one place could make sense, but up to you when you want to do that.

```suggestion
// GetIntervalFrom returns the minimum interval. 
// dsInterval is the string representation of data source min interval, if configured.
// queryInterval is the string representation of query interval (min interval), e.g. "10ms" or "10s".
// queryIntervalMS is a pre-calculated numeric representation of the query interval in milliseconds.
func GetIntervalFrom(dsInterval, queryInterval string, queryIntervalMS int64, defaultInterval time.Duration) (time.Duration, error) {
	if queryInterval == "" {
		if queryIntervalMS != 0 {
			return time.Duration(queryIntervalMS) * time.Millisecond, nil
		}
	}

	interval := queryInterval
	if queryInterval == "" && dsInterval != "" {
		interval = dsInterval
		
	}

	if interval == "" {
		return defaultInterval, nil
	}

	interval = strings.Replace(strings.Replace(interval, "<", "", 1), ">", "", 1)
	isPureNum, err := regexp.MatchString(`^\d+$`, interval)
	if err != nil {
		return time.Duration(0), err
	}
	if isPureNum {
		interval += "s"
	}
	parsedInterval, err := time.ParseDuration(interval)
	if err != nil {
		return time.Duration(0), err
	}

	return parsedInterval, nil
}
```